receiver/awscontainerinsightreceiver/internal/gpu/decorator.go

```diff
@@ -0,0 +1,141 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package gpu
+
+import (
+	"context"
+
+	ci "github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/containerinsight"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awscontainerinsightreceiver/internal/cadvisor/extractors"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awscontainerinsightreceiver/internal/stores"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.uber.org/zap"
+)
+
+const (
+	gpuUtil        = "DCGM_FI_DEV_GPU_UTIL"
+	gpuMemUtil     = "DCGM_FI_DEV_FB_USED_PERCENT"
+	gpuMemUsed     = "DCGM_FI_DEV_FB_USED"
+	gpuMemTotal    = "DCGM_FI_DEV_FB_TOTAL"
+	gpuTemperature = "DCGM_FI_DEV_GPU_TEMP"
+	gpuPowerDraw   = "DCGM_FI_DEV_POWER_USAGE"
+)
+
+var metricToUnit = map[string]string{
+	gpuUtil:        "Percent",
+	gpuMemUtil:     "Percent",
+	gpuMemUsed:     "Bytes",
+	gpuMemTotal:    "Bytes",
+	gpuTemperature: "None",
+	gpuPowerDraw:   "None",
+}
+
+// GPU decorator acts as an interceptor of metrics before the scraper sends them to the next designated consumer
+type decorateConsumer struct {
+	containerOrchestrator string
+	nextConsumer          consumer.Metrics
+	k8sDecorator          Decorator
+	logger                *zap.Logger
+}
+
+func (dc *decorateConsumer) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{
+		MutatesData: true,
+	}
+}
+
+func (dc *decorateConsumer) ConsumeMetrics(ctx context.Context, md pmetric.Metrics) error {
+	resourceTags := make(map[string]string)
+	rms := md.ResourceMetrics()
+	for i := 0; i < rms.Len(); i++ {
+		// get resource attributes
+		ras := rms.At(i).Resource().Attributes()
+		ras.Range(func(k string, v pcommon.Value) bool {
+			resourceTags[k] = v.AsString()
+			return true
+		})
+		ilms := rms.At(i).ScopeMetrics()
+		for j := 0; j < ilms.Len(); j++ {
+			ms := ilms.At(j).Metrics()
+			for k := 0; k < ms.Len(); k++ {
+				m := ms.At(k)
+				fields, tags := ci.ConvertToFieldsAndTags(m, dc.logger)
+				// copy down resource metrics only when it's missing at datapoint
+				for rtk, rtv := range resourceTags {
+					if _, ok := tags[rtk]; !ok {
+						tags[rtk] = rtv
+					}
+				}
```

Comment:
I think you mean "copy down resource attributes". Can you explain why we're doing this in the receiver?

Reply:
Yeah, this doesn't really add any value. removed.

If you are asking why we are adding attributes in the receiver? It was modeled similar to CI metric generation. There is no strong reason this should be done at the receiver.